x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/IDGenerator.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.dataframe.transforms;
+
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.BytesRefBuilder;
+import org.elasticsearch.common.Numbers;
+import org.elasticsearch.common.hash.MurmurHash3;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+
+/**
+ * ID Generator for creating unique but deterministic document ids.
+ *
+ * uses MurmurHash with 128 bits
+ */
+public class IDGenerator {
+    private static final byte[] NULL_VALUE = "__NULL_VALUE__".getBytes(StandardCharsets.UTF_8);
+    private static final BytesRef DELIM = new BytesRef("$");
+    private static final long SEED = 19;
+
+    private final BytesRefBuilder buffer = new BytesRefBuilder();
+
+    public IDGenerator() {
+    }
+
+    /**
+     * Clear the internal buffer to reuse the generator
+     */
+    public void clear() {
+        buffer.clear();
+    }
+
+    /**
+     * Add a value to the generator
+     * @param value the value
+     */
+    public void add(Object value) {
+        byte[] v = getBytes(value);
+
+        buffer.append(v, 0, v.length);
+        buffer.append(DELIM);
+    }
+
+    /**
+     * Create a document id based on the input objects
+     *
+     * @return a document id as string
+     */
+    public String getID() {
```

Comment:
This would make usage quite complicated. Note that this is an internal class, it's only used in 1 place, so I do not see the danger of a usage bug.

Reply:
This class is generating deterministic IDs so by it's very nature it will be used in more than one place when calculating the IDs for update etc. The field order is important as it changes the ID, at the moment the caller must know that edge case and pre-sort. I think it would be simpler to pass the fully constructed doc `Map<String,Object>` and get an ID back.

Reply:
A fully constructed doc wouldn't work because it would not know which fields it should use. Only the groupBy field values are used for id generation.

What I could do is storing all `add's` in a `TreeMap`and move all the magic into `getID`.

Reply:
>This would make usage quite complicated.

I am not sure how passing the full parameters to a static method is more complicated than handling state in this object. You will just be pushing the state handling to somebody else, which seems preferable to me in this instance, and will prevent misuse. 

I am glad the idea of "clear" is gone.

Reply:
I meant the consumer side which needs to iterate over results and add the fields, it would need to collect them locally before pushing it into the generator. Actually that's exactly how it works now with the only difference it is collected at the generator side. This is due to order dependency.

Reply:
My take (will approve PR regardless), is pushing this gathering to the consumer side is explicit in the usage and is consequently enforceable. Assuming that the consumer will construct a new class, call `add` for its fields, and THEN call ID is implicit, and not enforceable via the abstraction. 

Given that nobody else is concerned about this, and that I have no other qualms about this PR being merged, 👍